### DISCUSSION TITLE
`Request.is_disconnected` could consume request body

### DISCUSSION THREAD

**Discussion by msdrigg** [encode/starlette#2215] [Potential Issue]
For some background, I got here looking for a way to cancel costly actions if the client disconnects. I checked the source of Request.is_disconnected:

```python
    async def is_disconnected(self) -> bool:
        if not self._is_disconnected:
            message: Message = {}

            # If message isn't immediately available, move on
            with anyio.CancelScope() as cs:
                cs.cancel()
                message = await self._receive()

            if message.get("type") == "http.disconnect":
                self._is_disconnected = True

        return self._is_disconnected
```

Here we see that `is_disconnected` calls `self._receive()` (which consumes a asgi message). I believe that this could possibly consume and discard a `http.request` message containing the request body. 

If this is actually a possible side effect, I believe that it should be avoided (or at least documented somewhere).

If this is not possible to fix, is there any way to listen for disconnection (ideally create an `Event` that signals disconnection) without consuming the request? I read the spec, but unless I buffer arbitrarily many messages from `receive`, I don't know how I would do this.

I am relatively new to async python web frameworks, so I may be missing something obvious (or asking in the right place), but I wanted to reach out to see if anyone here could help me

**Comment by msdrigg** [comment]
Linking https://github.com/tiangolo/fastapi/discussions/8805 for context

**Comment by aeb-dev** [comment]
We are also searching for a way to implement this

**Comment by hari01584** [comment]
Can you try and see if https://github.com/tiangolo/fastapi/discussions/11360 fix your problem?
